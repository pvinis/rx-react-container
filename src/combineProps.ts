import { Observable, Observer, of } from 'rxjs';
import { map } from 'rxjs/operators';

import { combineLatestObj } from './combineLatestObj';

type AnyObservablesObject = { [k: string]: Observable<any> };
type AnyObserversObject = { [k: string]: Observer<any> };

/**
 * Creates observable combining values from observables, observers(as callbacks) and plain object
 * resulting in Observable of properties to be rendered with react component.
 */
export function combineProps<
  ObservablesObject extends AnyObservablesObject,
  ObserversObject extends AnyObserversObject,
  OtherProps extends {}
>(
  observables?: ObservablesObject,
  observers?: ObserversObject,
  props?: OtherProps
): Observable<
  OtherProps &
    {
      [k in keyof ObservablesObject]: ObservablesObject[k] extends Observable<
        infer V
      >
        ? V
        : never;
    } &
    {
      [k in keyof ObserversObject]: ObserversObject[k] extends Observer<infer V>
        ? (value: V) => void
        : never;
    }
> {
  const baseProps: any = Object.assign({}, props);

  if (observers) {
    Object.keys(observers).forEach(key => {
      baseProps[key] = (value: any) => {
        observers[key].next(value);
      };
    });
  }

  if (observables && Object.keys(observables).length > 0) {
    return combineLatestObj(observables).pipe(
      map(newProps => Object.assign({}, baseProps, newProps))
    );
  }

  return of(baseProps);
}
