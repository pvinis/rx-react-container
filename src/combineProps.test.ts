import { BehaviorSubject, of } from 'rxjs';

import { take, tap, toArray } from 'rxjs/operators';

import { combineProps } from './combineProps';

describe('combineProps', () => {
  it('works correctly for no arguments', done => {
    combineProps()
      .pipe(toArray())
      .subscribe(v => {
        expect(v).toMatchSnapshot();
        done();
      });
  });

  it('works correctly for empty arguments', done => {
    combineProps({}, {}, {})
      .pipe(toArray())
      .subscribe(v => {
        expect(v).toMatchSnapshot();
        done();
      });
  });

  it('works correctly when props are passed', done => {
    combineProps({}, {}, { a: 1 })
      .pipe(toArray())
      .subscribe(v => {
        expect(v).toMatchSnapshot();
        done();
      });
  });

  it('works correctly for observables', done => {
    combineProps(
      {
        a: of(1),
        b: of(2),
      },
      {},
      {}
    )
      .pipe(toArray())
      .subscribe(v => {
        expect(v).toMatchSnapshot();
        done();
      });
  });

  it('works correctly observers', done => {
    const a$ = new BehaviorSubject(0);
    combineProps(
      {},
      {
        a: a$,
      },
      {}
    )
      .pipe(toArray())
      .subscribe(v => {
        expect(v).toMatchSnapshot();
        v[0].a(123);
        expect(a$.value).toBe(123);
        done();
      });
  });

  it('works for complete sample', done => {
    const b$ = new BehaviorSubject(0);
    combineProps(
      {
        a: of(1),
        b: b$,
      },
      {
        onB: b$,
      },
      {
        c: 1,
        d: 2,
      }
    )
      .pipe(
        tap(({ onB }) => setTimeout(onB, 1, 1)),
        take(2),
        toArray()
      )
      .subscribe(v => {
        expect(v).toMatchSnapshot();
        done();
      });
  });
});
