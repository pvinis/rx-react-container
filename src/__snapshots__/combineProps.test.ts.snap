// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`combineProps works correctly for empty arguments 1`] = `
Array [
  Object {},
]
`;

exports[`combineProps works correctly for no arguments 1`] = `
Array [
  Object {},
]
`;

exports[`combineProps works correctly for observables 1`] = `
Array [
  Object {
    "a": 1,
    "b": 2,
  },
]
`;

exports[`combineProps works correctly observers 1`] = `
Array [
  Object {
    "a": [Function],
  },
]
`;

exports[`combineProps works correctly when props are passed 1`] = `
Array [
  Object {
    "a": 1,
  },
]
`;

exports[`combineProps works for complete sample 1`] = `
Array [
  Object {
    "a": 1,
    "b": 0,
    "c": 1,
    "d": 2,
    "onB": [Function],
  },
  Object {
    "a": 1,
    "b": 1,
    "c": 1,
    "d": 2,
    "onB": [Function],
  },
]
`;
